### PR TITLE
feat(model) Add ContainerPath aspect model

### PR DESCRIPTION
**Pull request by chriscollins3456**
This adds the new `containerPath` aspect model. This will be used for our updated browse experience to show a different type of path.

Feel free to think more about the model and how it will be used and have a discussion here on this PR!

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by jjoyce0510 — APPROVED** [review]
LGTM - Thank you!

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPath.pdl
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/common/ContainerPath.pdl:24]: I know we discussed this previously, but are we _SURE_ that this modeling restriction won't get us into trouble?

Comment (jjoyce0510) [reply metadata-models/src/main/pegasus/com/linkedin/common/ContainerPath.pdl:24]: Is there any world in which it might?

**Comment by xiphl** [comment]
Is the eventual plan to show containers and datasets together in browse?

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPath.pdl
Comment (chriscollins3456) [reply metadata-models/src/main/pegasus/com/linkedin/common/ContainerPath.pdl:24]: hmm @jjoyce0510 do you have anything in mind? i'm definitely going to keep thinking about it before merging... and get approvals from at least one of the ingestion folks

File: metadata-models/src/main/resources/entity-registry.yml
Comment (hsheth2) [inline metadata-models/src/main/resources/entity-registry.yml:144]: the containerPath for a container should not include itself, right?

for root level containers, should be emit an empty containerPath?

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl
Comment (hsheth2) [inline metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl:8]:
```suggestion
   * The ID of the container path entry. This is what gets stored in the index after URL encoding.
```

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl
Comment (asikowitz) [inline metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl:11]: Could we ever want to use container paths to support a more custom "folder" type experience, where users can choose what path they want entities in, especially those that don't fit in containers? I think the lines are a little blurred because ContainerPaths aren't just for storing the path of containers for an entity, it's the path of containers for an entity + another path for entities that aren't stored in containers.

I know this is a significant change, but what do you think about making container paths be composed of _only_ containers (thus can be a list of urns), then having a second aspect like `CustomBrowsePath` that we use for the non-container paths?

File: metadata-models/src/main/resources/entity-registry.yml
Comment (chriscollins3456) [reply metadata-models/src/main/resources/entity-registry.yml:144]: that's right! that's what i would expect

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl
Comment (chriscollins3456) [reply metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl:8]: thank you

**Comment by chriscollins3456** [comment]
> Is the eventual plan to show containers and datasets together in browse?

@xiphl yes! we're working towards a new browse experience that's based on containers for datasets

File: metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl
Comment (chriscollins3456) [reply metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl:11]: hmm yeah i definitely get what you're saying. Thinking about implementation, I personally would rather not have to know whether a given entity has `containerPaths` or some other aspect in order for us to operate browse.

The other thing is that we're going to need to filter on this aspect across all entity types (when they click a group on the sidebar of browse) so having that custom logic to filter for one field on elastic or another may be frustrating.

Maybe we can come up with a better name other than `containerPaths`? however the word "container" is already generic enough - it's just a little overloaded since we have actual "container" entities.

what do you think?

Comment (asikowitz) [reply metadata-models/src/main/pegasus/com/linkedin/common/ContainerPathEntry.pdl:11]: I think the name is ok, maybe there's a better one but I don't feel too strongly about it. Thinking over it more, this seems reasonable enough to me!

**Review by asikowitz — APPROVED** [review]

**Review by jjoyce0510 — APPROVED** [review]

File: metadata-models/src/main/resources/entity-registry.yml
Comment (hsheth2) [reply metadata-models/src/main/resources/entity-registry.yml:144]: perfect!

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**
